modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/OnboardServiceImpl.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018-2019 Expedia Group, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expedia.adaptivealerting.modelservice.service;
+
+import com.expedia.adaptivealerting.modelservice.entity.Metric;
+import com.expedia.adaptivealerting.modelservice.entity.MetricTagMapping;
+import com.expedia.adaptivealerting.modelservice.entity.Tag;
+import com.expedia.adaptivealerting.modelservice.repo.ItemExistsException;
+import com.expedia.adaptivealerting.modelservice.repo.MetricRepository;
+import com.expedia.adaptivealerting.modelservice.repo.MetricTagMappingRepository;
+import com.expedia.adaptivealerting.modelservice.repo.TagRepository;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * OnboardServiceDetail Implementation
+ *
+ * @author tbahl
+ */
+@Service
+@Slf4j
+public class OnboardServiceImpl implements OnboardService {
+
+    @Autowired
+    private TagRepository tagRepository;
+
+
+    @Autowired
+    private MetricRepository metricRepository;
+
+    @Autowired
+    private MetricTagMappingRepository metricTagMappingRepository;
+
+    @Override
+    public Metric onboard(Metric metric) {
+        List<Tag> tagList = new ArrayList<>();
+        Metric metricEntry = null;
+        Tag tagEntry;
+        String hash = metric.getHash();
+        Metric m = metricRepository.findByHash(hash);
+        if (m != null) {
+            throw new ItemExistsException(m);
+        }
+        for (Iterator<Map.Entry<String, Object>> tagIterator = metric.getTags().entrySet().iterator(); tagIterator.hasNext(); ) {
+            Map.Entry<String, Object> entry = tagIterator.next();
+
+            metricRepository.save(metric);
+            metricEntry = metricRepository.findByHash(metric.getHash());
+            tagList = tagRepository.findByUkeyContainingAndUvalueContaining(entry.getKey(), (String) entry.getValue());
+
+            if (tagList.size() == 0) {
+                tagEntry = tagRepository.save(new Tag(entry.getKey(), (String) entry.getValue()));
+                metricTagMappingRepository.save(new MetricTagMapping(metricEntry, tagEntry));
+            } else { metricTagMappingRepository.save(new MetricTagMapping(metricEntry, tagList.get(0))); }
```

Comment:
See if you can skip 1 else here. Ideally, if the tag doesn't exist then we need to add an entry in the tag table. We need to add mapping irrespective of the fact whether the tag exists or not.